src/libstd/sys/unix/weak.rs

```diff
@@ -45,23 +45,22 @@ impl<F> Weak<F> {
         }
     }
 
-    pub fn get(&self) -> Option<&F> {
+    pub fn get(&self) -> Option<F> {
         assert_eq!(mem::size_of::<F>(), mem::size_of::<usize>());
         unsafe {
             if self.addr.load(Ordering::SeqCst) == 1 {
                 self.addr.store(fetch(self.name), Ordering::SeqCst);
             }
-            if self.addr.load(Ordering::SeqCst) == 0 {
-                None
-            } else {
-                mem::transmute::<&AtomicUsize, Option<&F>>(&self.addr)
```

Comment:
I suspect this old code may actually have been lurking UB, though I doubt it was a problem in practice. AIUI the memory const-ness and aliasing rules are different for `AtomicUsize` (with `UnsafeCell`) versus any plain type like this `F` function pointer.
cc @RalfJung

Reply:
I'm afraid I am not familiar with this code. Even disregarding Stacked Borrows, this looks like it performs a non-atomic read of an `AtomicUsize`. If there is any chance of a data race, that's UB.

Whether that is actually the case here I cannot tell at a quick glance, someone would have to analyze this code more carefully.

Reply:
The race that I see is in the initialization sequence:
1. atomic load sees 1 (needs init)
2. atomic store the function pointer as usize
3. atomic load sees non-zero
4. transmute `&AtomicUsize` to `Option<&F>`
5. plain load to read and call the function

Two threads could see that we need init in step 1, and then one thread's store in step 2 could race with the other's load in step 5.

After initialization, the value is only read, so I guess there's no further chance to race.

@RalfJung The more general question I was trying to get at is perhaps something for unsafe code guidelines. Should it be *legal* to have a reference to an `UnsafeCell` which is aliased with a reference to a plain type? i.e. the former is a reference that the compiler knows to have interior mutability, and the latter is a reference that the compiler thinks is fully immutable. Or perhaps this is _only_ considered a problem if you actually manifest a race, writing the former while reading the latter.

Reply:
(The new code in this PR should be fine, always using atomic access.)

Reply:
> Should it be _legal_ to have a reference to an `UnsafeCell` which is aliased with a reference to a plain type? i.e. the former is a reference that the compiler knows to have interior mutability, and the latter is a reference that the compiler thinks is fully immutable. Or perhaps this is _only_ considered a problem if you actually manifest a race, writing the former while reading the latter.

This is actually possible in safe code, you can e.g. have a reference to a `RefCell` and also a reference to its interior.

The important point is that between the "plainly typed" reference (the one that does not "expect" mutation) being created and it being used the last time, the pointee must not be written to. As a "use" here counts not just actually accessing memory, but also reborrowing, copying, passing to a function -- anything the reference is involved in.

Reply:
Ah, right, and `RefCell` has its own protection that `borrow_mut()` is not allowed while any "plain" reference via `borrow()` is alive.

Then I think the code here did have a race, but the general situation is OK. Thanks!